contracts/core/mpl-rewards/v1/test/helper/Rewards.sol

```diff
@@ -0,0 +1,740 @@
+
+        assertEq(stakeToken.custodyAllowance(address(fay), address(custodian)), 0);  // Custodian allowance has been reduced
+        assertEq(stakeToken.totalCustodyAllowance(address(fay)),                0);  // Total custody allowance has been reduced, giving Fay access to funds again
+    }
+
+    /****************************/
+    /*** LP functions testing ***/
+    /****************************/
+    function stake_test(bool isPfdtStakeToken, uint256 amt, uint256 stakeAmt, IStakeToken stakeToken) public {
+        uint256 start = block.timestamp;
+
+        if (isPfdtStakeToken) {
+            mintFundsAndDepositIntoPool(fay, pool1, amt * USD, amt * USD);
+            pat.setLockupPeriod(address(pool1), 0);
+        } else {
+            setUpForStakeLocker(amt, sam, fay);
+        }
+
+        checkDepositOrStakeDate(isPfdtStakeToken, start, stakeToken, fay);
+
+        amt *= WAD;
+
+        assertEq(stakeToken.balanceOf(address(fay)),          amt);
+        assertEq(mplRewards.balanceOf(address(fay)),            0);
+        assertEq(mplRewards.totalSupply(),                      0);
+
+        hevm.warp(start + 1 days);  // Warp to ensure no effect on depositDates
+
+        assertTrue(!fay.try_stake(stakeAmt * WAD));  // Can't stake before approval
+
+        fay.increaseCustodyAllowance(address(mplRewards), stakeAmt * WAD);
+
+        assertTrue(!fay.try_stake(0));          // Can't stake zero
+        assertTrue( fay.try_stake(stakeAmt * WAD));  // Can stake after approval
+
+        assertEq(stakeToken.balanceOf(address(fay)),     amt);  // PoolFDT balance doesn't change
+        assertEq(mplRewards.balanceOf(address(fay)),     100 * WAD);
+        assertEq(mplRewards.totalSupply(),               100 * WAD);
+
+        checkDepositOrStakeDate(isPfdtStakeToken, start, stakeToken, fay);
+    }
+
+    function withdraw_test(bool isPfdtStakeToken, uint256 amt, uint256 stakeAmt, IStakeToken stakeToken) public {
+        uint256 start = block.timestamp;
+
+        if (isPfdtStakeToken) {
+            mintFundsAndDepositIntoPool(fay, pool1, amt * USD, amt * USD);
+            assertEq(stakeToken.balanceOf(address(fay)), amt * WAD);
+            pat.setLockupPeriod(address(pool1), 0);
+        } else {
+            setUpForStakeLocker(amt, sam, fay);
+        }
+
+        amt      *= WAD;
+        stakeAmt *= WAD;
+
+        fay.increaseCustodyAllowance(address(mplRewards), stakeAmt);
+        fay.stake(stakeAmt);
+
+        hevm.warp(start + 1 days);  // Warp to ensure no effect on depositDates
+
+        checkDepositOrStakeDate(isPfdtStakeToken, start, stakeToken, fay);
+
+        assertEq(stakeToken.balanceOf(address(fay)),      amt);  // FDT balance doesn't change
+        assertEq(mplRewards.balanceOf(address(fay)), stakeAmt);
+        assertEq(mplRewards.totalSupply(),           stakeAmt);
```

Comment:
We should add an assertion here for `totalCustodyAllowance`, asserting that it goes back down after withdraw